### PR TITLE
Introduce examples for pipeline api.

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
examples for #584

### CODE REVIEW

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/SimpleTrainAndPredict.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        /// (e.g., the prediction does not happen over a file as it did during training).
+        /// </summary>
+        [Fact]
+        void SimpleTrainAndPredict()
+        {
+            var dataPath = GetDataPath(SentimentDataPath);
+            var testDataPath = GetDataPath(SentimentDataPath);
+            var pipeline = new LearningPipeline();
+
+            pipeline.Add(MakeSentimentTextLoader(dataPath));
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/SimpleTrainAndPredict.cs]: >MakeSentimentTextLoader [](start = 25, length = 23)

can you not use TextLoader<IrisData> somehow?

Comment (Ivanidzo4ka) [reply test/Microsoft.ML.Tests/Scenarios/PipelineApi/SimpleTrainAndPredict.cs]: You want text loader with arguments? Or what?

---
In reply to: [209770697](https://github.com/dotnet/machinelearning/pull/677#discussion_r209770697) [](ancestors = 209770697)

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/CrossValidation.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        /// and trainer pipeline, and it will handle (1) splitting up the data, (2) training the separate
+        /// pipelines on in-fold data, (3) scoring on the out-fold data, (4) returning the set of
+        /// evaluations and optionally trained pipes. (People always want metrics out of xfold,
+        /// they sometimes want the actual models too.)
+        /// </summary>
+        [Fact]
+        void CrossValidation()
+        {
+            var dataPath = GetDataPath(SentimentDataPath);
+            var testDataPath = GetDataPath(SentimentDataPath);
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/CrossValidation.cs]: >testDataPath [](start = 16, length = 12)

remove line

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/CrossValidation.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            var pipeline = new LearningPipeline();
+
+            pipeline.Add(MakeSentimentTextLoader(dataPath));
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/CrossValidation.cs]: >Add [](start = 21, length = 3)

TextLoader<SentimentData> ?

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/Metacomponents.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        void Metacomponents()
+        {
+            var dataPath = GetDataPath(IrisDataPath);
+            var pipeline = new LearningPipeline(seed: 1, conc: 1);
+            pipeline.Add(new TextLoader(dataPath).CreateFrom<IrisData>(useHeader: false));
+            pipeline.Add(new Dictionarizer(new[] { "Label" }));
+            pipeline.Add(new ColumnConcatenator(outputColumn: "Features",
+                "SepalLength", "SepalWidth", "PetalLength", "PetalWidth"));
+
+            pipeline.Add(OneVersusAll.With(new StochasticDualCoordinateAscentBinaryClassifier()));
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/Metacomponents.cs]: >With [](start = 38, length = 4)

does it accept multiclass?

Comment (Zruty0) [reply test/Microsoft.ML.Tests/Scenarios/PipelineApi/Metacomponents.cs]: write a comment that this is checked at Train time, so technically this example is not possible with LP

---
In reply to: [210040514](https://github.com/dotnet/machinelearning/pull/677#discussion_r210040514) [](ancestors = 210040514)

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/MultithreadedPrediction.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            var threadEngine = new ThreadLocalBasedPredictionEngine(modelName);
+            var poolEngine = new PoolBasedPredictionEngine(modelName);
+            Parallel.ForEach(collection, (input) => threadEngine.Predict(input));
+            Parallel.ForEach(collection, (input) => poolEngine.Predict(input));
+        }
+
+        /// <summary>
+        /// This is a trivial implementation of a thread-safe prediction engine is just guarded by a lock.
+        /// </summary>
+        private sealed class LockBasedPredictionEngine
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/MultithreadedPrediction.cs]: >class [](start = 23, length = 5)

remove custom classes

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/MultithreadedPrediction.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+            pipeline.Add(MakeSentimentTextTransform());
+
+            pipeline.Add(new FastTreeBinaryClassifier() { NumLeaves = 5, NumTrees = 5, MinDocumentsInLeafs = 2 });
+
+            pipeline.Add(new PredictedLabelColumnOriginalValueConverter() { PredictedLabelColumn = "PredictedLabel" });
+            var model = pipeline.Train<SentimentData, SentimentPrediction>();
+            var modelName = "multithreadModel.zip";
+            DeleteOutputPath(modelName);
+            model.WriteAsync(modelName);
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/MultithreadedPrediction.cs]: >model [](start = 12, length = 5)

await?

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/SimpleTrainAndPredict.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+            pipeline.Add(MakeSentimentTextTransform());
+
+            pipeline.Add(new FastTreeBinaryClassifier() { NumLeaves = 5, NumTrees = 5, MinDocumentsInLeafs = 2 });
+
+            pipeline.Add(new PredictedLabelColumnOriginalValueConverter() { PredictedLabelColumn = "PredictedLabel" });
+            var model = pipeline.Train<SentimentData, SentimentPrediction>();
+            var testData = MakeSentimentTextLoader(testDataPath);
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/SimpleTrainAndPredict.cs]: >MakeSentimentTextLoader [](start = 27, length = 23)

yuck

Comment (Zruty0) [reply test/Microsoft.ML.Tests/Scenarios/PipelineApi/SimpleTrainAndPredict.cs]: rename testData to testPipe

---
In reply to: [210042025](https://github.com/dotnet/machinelearning/pull/677#discussion_r210042025) [](ancestors = 210042025)

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            var model = pipeline.Train<SentimentData, SentimentPrediction>();
+
+            var modelName = "trainSaveAndPredictdModel.zip";
+            DeleteOutputPath(modelName);
+            await model.WriteAsync(modelName);
+
+            var loadedModel = await PredictionModel.ReadAsync<SentimentData, SentimentPrediction>(modelName);
+            var testData = MakeSentimentTextLoader(testDataPath);
+            var evaluator = new BinaryClassificationEvaluator();
+            var metrics = evaluator.Evaluate(loadedModel, testData);
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs]: remove from this test

Comment (Zruty0) [reply test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs]: Move to evaluation

---
In reply to: [210042424](https://github.com/dotnet/machinelearning/pull/677#discussion_r210042424) [](ancestors = 210042424)

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            var singlePrediction = loadedModel.Predict(new SentimentData() { SentimentText = "Not big fan of this." });
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs]: >SentimentData [](start = 59, length = 13)

try to do the same with Iris, and check if loading a model with `IrisDataNoLabel` works, but `Predict` fails. If that's the case, then there's no decomposable piope.

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/Evaluation.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+            pipeline.Add(MakeSentimentTextTransform());
+            pipeline.Add(new FastTreeBinaryClassifier() { NumLeaves = 5, NumTrees = 5, MinDocumentsInLeafs = 2 });
+            pipeline.Add(new PredictedLabelColumnOriginalValueConverter() { PredictedLabelColumn = "PredictedLabel" });
+            var model = pipeline.Train<SentimentData, SentimentPrediction>();
+            var testLearningPipelineItem = new TextLoader(testDataPath).CreateFrom<SentimentData>();
+            var evaluator = new BinaryClassificationEvaluator();
+            var metrics = evaluator.Evaluate(model, testLearningPipelineItem);
+
+            var singlePrediction = model.Predict(new SentimentData() { SentimentText = "Not big fan of this." });
+            Assert.True(singlePrediction.Sentiment);
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/Evaluation.cs]: remove

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/Metacomponents.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        void Metacomponents()
+        {
+            var dataPath = GetDataPath(IrisDataPath);
+            var pipeline = new LearningPipeline(seed: 1, conc: 1);
+            pipeline.Add(new TextLoader(dataPath).CreateFrom<IrisData>(useHeader: false));
+            pipeline.Add(new Dictionarizer(new[] { "Label" }));
+            pipeline.Add(new ColumnConcatenator(outputColumn: "Features",
+                "SepalLength", "SepalWidth", "PetalLength", "PetalWidth"));
+
+            // this will throw exception during training time if you specify any other than binary classifier.
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/Metacomponents.cs]: >t [](start = 15, length = 1)

T

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs (hunk 9 above)
Comment (Ivanidzo4ka) [reply test/Microsoft.ML.Tests/Scenarios/PipelineApi/TrainSaveModelAndPredict.cs]: 'Source column 'Label' is required but not found'


---
In reply to: [210044162](https://github.com/dotnet/machinelearning/pull/677#discussion_r210044162) [](ancestors = 210044162)

**Comment by Ivanidzo4ka** [comment]
@codemzs  @eerhardt  I'm trying to introduce examples for #584, and right now I have
Simple train and predict
Multi-threaded prediction
Train, save/load model, predict
Evaluation
Cross-validation
Meta-components
and I think it's impossible to implement in LearningPipeline following ones:
Train with validation set
Train with initial predictor
Auto-normalization and caching
File-based saving of data
Decomposable train and predict
Reconfigurable predictions
Introspective training
Exporting models
Visibility
Extensibility

Can you confirm or tell me I'm wrong for some cases.

**Comment by eerhardt** [comment]
I can only answer a few. Maybe @codemzs can answer the others.

> Train with validation set
Train with initial predictor

I don't see any Entry Point for these scenarios, let alone a LearningPipeline API...

> Exporting models

Isn't this `PredictionModel.WriteAsync`?

File: test/Microsoft.ML.Tests/Scenarios/PipelineApi/MultithreadedPrediction.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            pipeline.Add(new PredictedLabelColumnOriginalValueConverter() { PredictedLabelColumn = "PredictedLabel" });
+            var model = pipeline.Train<SentimentData, SentimentPrediction>();
+            var collection = new List<SentimentData>();
+            int numExamples = 100;
+            for (int i = 0; i < numExamples; i++)
+                collection.Add(new SentimentData() { SentimentText = "Let's predict this one!" });
+
+            Parallel.ForEach(collection, (input) =>
+            {
+                lock (model)
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Scenarios/PipelineApi/MultithreadedPrediction.cs]: I would put a comment here why we would need to lock.

**Pull request merged**